Counterfactual predictions

michaellevy opened the pull request:
@taylorlarsen  Could you review this in the next day or two? I know you have a million things on your plate right now, but the HAS-class textbook needs to go to the printers on Monday, and I'd like to include this function in it. I'm going to start working on plot and print methods for this in the meantime.

The name of the function needs to change: there is a `stats::simulate` function and we don't want that namespace conflict. So, if you have any ideas, feel free to suggest. 

There are a lot of different levels of customization. By default, it pulls variable importance (or coefficients from`interpret` if and only if glm is the only model present) and varies the `vary`-most important variables (default = 4). For categorical variables that make that cut, we use all the values in the training data; for numerics we use the 0, 25, 50, 75, and 100th percentile values from the training data. Those can be customized via the `characters` and `numerics` arguments, respectively. `characters` just lets you limit the number of categories to use (always from most to least common in the training data); `numerics` lets you choose the number of (evenly spaced) quantiles to use (e.g. `numerics = 3` will use the min, median, and max), or the quantiles themselves (e.g. `numerics = c(0, .5, 1)` does the same).

You can also choose which variables to vary in a variety of ways. If you pass `vary` an integer, you get the `vary`-most important variables. If you pass it a vector of integers, you get those most important variables (you asked for this... so you could do `vary = 1:3` for the top three, then `vary = 4:6` for the next three). You can also specify which variables to vary (e.g. `vary = c("weight_class", "age")`) or the values to use (e.g. `vary = list(weight_class = c("underweight", "overweight"), age = seq(20, 80, 5))`).

The values of non-varying variables to use are determined by the `hold` argument. There are two ways to specify this: either with functions to apply to the columns in the training data, or with values to use, which provides an easy way to run simulations for a particular patient. The default is to use the median value for numerics and the modal value for categories. You asked about being able to use the mean; to do that you'd specify `hold = list(numerics = mean, characters = Mode)`. To use a particular patient's values, just pass that row of the training dataframe (see the last example in the docs). Or if you want to pass custom values, you can pass in a list with names of variables and single values for each.

michaellevy commented:
Thanks @taylorlarsen. Plots soon, but just wanted to mention this super-secret, uber-hacky way to get just one variable that isn't the most important: use `vary = c(2, 2.1)` to get just the second.

michaellevy commented:
To do:

- [ ] tests for plot
- [ ] tests for error and warning messages in `simulate`
- [ ] rename `simulate`
- [ ] warn or slightly adjust y when lines are entirely overlapping (happens when a dummy variable isn't used by the model and its parent variable is mapped to color)
- [ ] Plot appearance customization (font_size, line_width, etc.) 
- [ ] New argument for strip text size since it often overflows
- [ ] Argument (logical) to rotate x-axis text to vertical
- [ ] Examples
- [ ] Improve `simulate` docs per Taylor's comments

mmastand commented:
Heart failure readmissions look pretty bad with the defaults because of big outliers. It either needs a log scale, or could use the 5th, 25th, 50th, 75th, 95th percentile?

![image](https://user-images.githubusercontent.com/15145347/44366120-f358f100-a488-11e8-8999-491ac599b1f7.png)

michaellevy commented:
@mmastand I'm not sure what looks bad about this to you. I think it's pretty informative: HxOfExcess being very small is a powerful negative predictor, and for really high HxOfEDACExposure, risk is linear across HxOfExcess while for lower HxOfEDACExposure it's a threshold response.

Where are you wanting a log scale here? We could add an argument like `log = c("x", "y", "color")`. I'd also be okay with making the facet defaults 5/25/50/75/95 instead of 0/25/50/75/100.

mmastand commented:
Bad is probably the wrong word. I mean that it's hard to tell what's actually happening in the lower end of the scale since the 100th percentile washes it out so strongly.

mmastand commented:
It also makes it hard to see what values are being used for HxOfEDAC on the color scale. Would that look better as a factor?

mmastand commented:
The following caused R to hang. I made it through predicting counterfactuals the first 2 times, but adding the `hold` argument broke it.
```
Deidentified_HF <- read_csv("Deidentified HF.csv")
d <- purrr::map_df(Deidentified_HF, function(x){
  x <- replace(x, x == "NULL", NA)
})

to_num <- c("HighestDailyGlucoseNBR", "HxHFAppointmentCNT", "AverageBNPVAL",
            "OralDrugsCNT", "EjectionFractionVAL", "BUNResultVAL",
            "CreatinineResultVAL", "HighIntensityEncounterCNT", "NoShowLWBSHFAppointmentCNT")
d[to_num] <- map_df(d[to_num], as.numeric)
d$PatientZipCD <- NULL

d_clean <- prep_data(d,
                     PredictThisPatientFLG, HospitalAccountID,
                     outcome = ThirtyDayReadmissionFLG,
                     remove_near_zero_variance = FALSE)
m <- 
  d_clean %>% 
  flash_models(outcome = ThirtyDayReadmissionFLG, n_folds = 5, models = c("glm"))
m

predict_counterfactuals(m)

predict_counterfactuals(m, vary = c("HxOfEDACExposureDaysNBR", "HxAppointmentCNT"))

patient321 <- dplyr::filter(d, HospitalAccountID == 321)
patient321
predict_counterfactuals(m, hold = patient321) %>% plot()
```
(data via slack)

michaellevy commented:
I like `explore`. @taylorlarsen you good with that?

taylorlarsen commented:
@michaellevy that will fit well for this function.

michaellevy commented:
@mmastand great feedback, thank you. Some thoughts as I update based on your comments:

- That example doesn't hang, it just takes a long time because AverageMSDRGWeightNBR is a character for some reason with 1811 unique values, leading to a large `expand.grid` output and a ggplot with 1811 levels across the x-axis 😨. It's a good tripping point for us to watch out for. I changed the default number of levels to use in categorical variables from Inf to 10. 

- "The 5x5 default grid felt like a ton of information..." I agree, and it's mostly information of secondary importance. I changed default `n_use` to 2 so you just get features on x and color by default, but with defaults you also get the message about aggregation across other varying features, and I added a suggestion to turn up `n_use` to see more features' effects.

- "Percentiles with big outliers can wash out interactions in other parts of the range." Two changes made here: we now convert numerics mapped to color to factors. I don't love giving up the quantitative information there, but it at least lets the user see what numbers are being used and which line corresponds to which number. I also changed the default numeric breakdown to use the 5th and 95th percentiles as endpoints instead of min and max.

- "When graph titles have a numeric, they can end up with an age of say, 58.340957804967273 years. Round?" I spent a lot of time looking for a good solution to this and didn't find one. The problem with `round` is that .0000123 becomes 0. The problem with `signif` is that 1234 becomes 1230. I agree though that the long decimals are awful. I added a `signif` call with an argument specifying the number of digits so that you can turn it off if, say, you have a variable where all the values are in [100,000, 200,000]
